src/DeltaLake/Kernel/Arrow/Extensions/ArrowContextExtensions.cs

```diff
@@ -21,29 +25,176 @@ namespace DeltaLake.Kernel.Arrow.Extensions
+
+                    Schema schemaWithPartition = AddPartitionColumnsToSchema(schemaWithoutPartition, partitionsPtr);
+                    RecordBatch batchWithPartition = AddPartitionColumnsToRecordBatch(batchWithoutPartition, partitionsPtr);
+
+                    recordBatches.Add(batchWithPartition);
+                    schemas.Add(schemaWithPartition);
+                }
             }
-            List<RecordBatch> recordBatches = new(context.NumBatches);
-            for (int i = 0; i < context.NumBatches; i++) recordBatches.Add(*context.Batches[i]);
-            return recordBatches;
+
+            for (int i = 1; i < schemas.Count; i++)
+            {
+                if (schemas[i] == schemas[0]) throw new InvalidOperationException($"All schemas must be the same in - got {i}th {schemas[i]} != 0th {schemas[0]}");
+            }
+
+            return (schemas[0], recordBatches);
         }
+
+        #endregion Extension Methods
+
+        #region Private Methods
+
+        private static void ValidateContext(this ArrowContext context)
+        {
+            if (context.NumBatches == 0)
+            {
+                throw new InvalidOperationException("Arrow Context must contain at least one RecordBatch");
+            }
+        }
+
+#pragma warning disable CA1859, IDE0060 // Although we're not using partitionValue right now, it will be used when Kernel supports reporting Arrow Schema
+        private static unsafe IArrowType DeterminePartitionColumnType(string colName, string colValue)
+        {
+            // Currently, there's no way to determine the type of the partition,
+            // because the Kernel always represents partition values as strings in CStringMap.
+            //
+            // We have a request with Kernel team here to get back the Arrow Schema from
+            // the Delta Transaction Log:
+            //
+            // >>> https://delta-users.slack.com/archives/C04TRPG3LHZ/p1728178727958499
+            // >>> https://delta-users.slack.com/archives/C04TRPG3LHZ/p1728001059452499?thread_ts=1727999835.930339&cid=C04TRPG3LHZ
+            //
+            return StringType.Default;
+        }
+#pragma warning restore CA1859, IDE0060
+
+        private static unsafe Schema AddPartitionColumnsToSchema(Apache.Arrow.Schema originalSchema, ParquetStringPartitions* partitionsPtr)
+        {
+            Apache.Arrow.Schema.Builder schemaBuilder = new();
+            foreach (Field field in originalSchema.FieldsList)
+            {
+                schemaBuilder = schemaBuilder.Field(field);
+            }
+
+            for (int i = 0; i < partitionsPtr->Len; i++)
+            {
+#pragma warning disable CS8600, CS8604 // If Kernel sends us back null pointers, we are in trouble anyway
+                string colName = Marshal.PtrToStringAnsi((IntPtr)partitionsPtr->ColNames[i]);
+                string colValue = Marshal.PtrToStringAnsi((IntPtr)partitionsPtr->ColValues[i]);

```

Comment:
These (and the ones in the next method) should almost certainly be PtrToStringUtf8

Reply:
Yessir, made changes to both now, also updated some other diffs in this PR.